Versioning.md

```diff
@@ -0,0 +1,38 @@
+
+`@internal` APIs are never intended to be used outside of the iTwin.js library itself. All other APIs are intended to ultimately become `@public` through the following process:
+
+- A highly experimental API may start out tagged as `@alpha`, indicating it is not yet mature enough for users to try out. This stage is usually skipped.
+- A new API, or one that has been promoted from `@alpha`, is tagged as `@beta` to invite users to experiment with it and provide feedback. The API may change in response to feedback from one release to the next.
+  - Note: extension APIs are tagged as `@preview` instead of `@beta`.
+- A `@beta` API that has been sufficiently tested is promoted to `@public` to indicate that it is now subject to [stability guarantees](#package-versioning-policy).
+- A `@public` API may be tagged as `@deprecated` indicating that it may later be removed according to the [deprecation policy](#api-deprecation-policy).
+
+Occasionally, a `@beta` or `@alpha` API may be abandoned rather than being promoted. Not all experiments are successful.
+
+In practice, `@beta` and `@public` APIs are typically designed in a way that promotes flexibility in the future - e.g., by accepting their arguments as "options" object types. This reduces the risk of breaking changes and deprecation.
+
+## Package versioning policy
+
+The individual packages that constitute the iTwin.js libraries use [semantic versioning](https://semver.org/) to convey information about how their public APIs change over time.
+
+- Patch releases occur frequently to fix bugs or security issues and have no impact on the public API.
+- Minor releases occur less frequently than patch releases. They may introduce new APIs, but never break any existing `@public` APIs.
+- Major releases occur only as frequently as necessary. They can include breaking changes to `@public` APIs, but only according to the [deprecation policy](#api-deprecation-policy).
+
+## API deprecation policy
+
+Most `@public` APIs are intended to remain stable indefinitely, but occasionally a breaking change is required to fix a bug, introduce new functionality, or address performance problems. The deprecation policy ensures that such changes are introduced deliberately and predictably to help users plan ahead.
+
+Before an API can be removed it must first be marked as `@deprecated` in version `N` of the package. It must remain stable for the duration of version `N+1`. Finally, in version `N+2` it *may* be removed. This means that, for example, an API deprecated in version 3.1 of a package cannot be removed until version 5.0 at the earliest.
```

Comment:
If we're going to begin enforcing this in V4, perhaps we should use 4.1 and 6.0 in this example.

Reply:
Do you intend to remove ui-related APIs in 4.x that were deprecated in 3.x?

Reply:
would it be better to just say `major version N` ? and not provide an explicit example?

Reply:
I would like to make it crystal clear to users when they can expect this policy to start being enforced, hence my question to @NancyMcCallB.

Reply:
@pmconne yes there is a plan to do it. The extent to which we do is yet to be determined. We're taking a look at the APIs now.

Also, we should talk with @grigasp I know he had some APIs he wanted to remove that are deprecated and not sure how long they've been that way.

Reply:
I agree with @NancyMcCallB, we should only start enforcing as of 4.x and onward. We were not clear previously.

However, this does not mean that we should go and add deprecated to everything we'd like to remove in 4.0 right now. We still need to avoid breaking things unnecessarily that will impact people.

Reply:
> We still need to avoid breaking things unnecessarily that will impact people.

Damn straight.

> this does not mean that we should go and add deprecated to everything we'd like to remove in 4.0

Maybe I misunderstand but if you are planning to remove something in 4.0, please deprecate it prior to that.

Reply:
@pmconne  sorry, wasn't clear. I meant that we should avoid adding anything else as deprecated in 3.x that we plan to remove in 4.0, i.e. we should be done trying to remove things in 4.0 that aren't already marked.

If you add deprecated now then you will need to remove it at the earliest in 5.0. Let me get the list together and see if that's the case. We should track all of this for the release.

Reply:
We have almost no deprecated APIs that we carried over from 2.x - all deprecated ones we currently have got deprecated in the 3.x timeframe. As much as I'd like to get rid of all of them, there's no strong reasons to remove them in 4.x except for one case - the APIs related to the `Table` component, which is going to be removed completely and there's no way for us to keep anything related to it.

Also, I'd like to suggest a couple of things documentation wise:

- When adding `@deprecated` tag, we should also specify the version it was deprecated in. Knowing that an API may be removed in version `N+1` doesn't help if you don't know the value of `N`. This can be found using `git`, but is really inconvenient.

- We should be clear on what we consider a breaking change. Here're a couple types of breaking changes (technically) that I believe we're making from time to time in our non-major releases:

  - Adding optional attributes to interfaces that are used as data structures. In theory, [every change to a public interface is a breaking change](https://log.havrlant.cz/interface-breaking-change/). I think such breaking changes should be allowed even in patches.
 
  - Upgrading peer dependency version even if it's just a patch release change. In theory, peer dependencies are considered part of the public API. See more details [here](https://github.com/semver/semver/issues/502). At the moment we have the packages all locked at the same version - increasing patch version of `core-bentley`, which is a peer dependency of nearly all other package in the repo, doesn't cause other packages' major version change. And that's technically violating the rule... I think this is going to become a bigger problem when UI and some Presentation packages move out of the core repo and start being versioned separately. Increasing major version of `presentation-components` with every patch release of `presentation-common` would cause a huge number of major releases, which we want to avoid. Either we have to keep violating the rule, or need to convert peer deps to regular deps by duplicating some of the interfaces.

Reply:
> Adding optional attributes to interfaces that are used as data structures. In theory, [every change to a public interface is a breaking change](https://log.havrlant.cz/interface-breaking-change/). I think this specific kind of breaking change should be allowed even in patches.

@grigasp , definitely agree with this one. I'd say this is fine as long as the API is still backwards compatible.

> Upgrading peer dependency version even if it's just a patch release change

Good point on this one, it is definitely confusing especially as we get more packages in separate places. We already handle this two different ways today when considering all of iTwin.js. The first way is what you mentioned for the itwinjs-core packages, every release we bump the version of peer dependencies for every other package in the repo (i.e. if they're in a lockstep release policy).  The second way, any time the minimum of a peer dependency is changed we require a major version bump to change as we consider that a breaking change (e.g. this is how we handle the itwinjs-core versions in the iTwin Viewer, every time we update the minimum core pkg versions we make a major version bump of the iTwin Viewer packages).

The general, unwritten, rule that we have been following is that anything that is within a lockstep release policy (i.e. itwinjs-core packages) can freely move up their peer dependency as long as that peer dependency is also within the lockstep policy. This ensures consistent versions and since the packages are all released exactly at the same time you need to do this action once to update them all.

When a package is outside of the release policy of it's peer dependency, it must follow the "second way". The main reason is that with them being on a separate version policy, and released independently, the coordination between the two packages is much more loose.  This does normally force the package that has a peer dependency to support a range of versions most times and I do always suggest when you're breaking this boundary you should always be using our "public" API. This is how we handle everything outside of the itwinjs-core repo.

So as an example, with the AppUI packages going to be removed from both the repo and the lockstep versioning of the rest of itwinjs-core, those packages will no longer be able to update the minimum of itwinjs core packages freely and instead be required to follow the "second way".  However, since they are moving into a repo of their own with its own lockstep versioning all of the AppUI packages will be released in lockstep, therefore subject to the "first way".

note: ideally we'd have independence between all of our packages and not require such a tight coupling but with the way we're architected and our extensive use of statics and the objects as parameters to methods it forces us to use peer dependencies quite a bit. We're in essence providing a "set" of packages in some cases rather than individual ones.

Reply:
Changed to 4.1.